storage/storage_image.go

```diff
@@ -137,10 +137,9 @@ func (s *storageImageDestination) putBlob(stream io.Reader, blobinfo types.BlobI
 		Digest: "",
 		Size:   -1,
 	}
-	// Try to read an initial snippet of the blob.
-	header := make([]byte, 10240)
-	n, err := stream.Read(header)
-	if err != nil && err != io.EOF {
+	buf := [archive.HeaderSize]byte{}
+	n, err := io.ReadAtLeast(stream, buf[:], len(buf))
+	if err != nil && err != io.EOF && err != io.ErrUnexpectedEOF {
```

Comment:
In general, `ErrUnexpectedEOF` would indeed be a problem. But in this case we don’t yet know whether there is a fixed-size block of any kind in the input: the input may be an uncompressed tar (which does have to start with a 512-byte block), a compressed file (where the initial header may be much smaller), or an uncompressed JSON (which can quite plausibly be smaller than 512 bytes).

The only reason to use `io.ReadAtLeast` here is that we do read all of `archive.HeaderSize` _if_ the input is that large, even if a single `stream.Read()` would return a smaller single chunk of data.